### ISSUE TITLE
patch: Interactive frozen script goes to background on SIGINT

### ISSUE THREAD

**Issue by pyinstaller-tickets-migration** [pyinstaller/pyinstaller#208]
**Original date**: 2010/09/20

Hi,

I have frozen a python program that implements a user CLI.
Normally, it dies catch KeyboardInterrupt, but when frozen this does not works anymore as expected. Setting a SIGINT handler in my python code shows that the signal indeed gets to it, but no matter how I handle the signal at the python level, it seems that the froxen process goes to the background. I use readline, and am unsure if this is connected, but when the process is in the background, I still get garbage on the terminal (my prompt it seems), like if the std in/out were not completely disconnected.

I am using a onefile frozen binary, with pyinstaller 1.4.

When I apply the following patch to make sure the wrapper pyinstaller code does not catch signals anymore, everything works as expected. The patch is next. What do you think ? Is there some problem doing that ? Does my patch makes sense ? At least it solves the problem for me....

```
--- pyinstaller-1.4/source/linux/main.c 2009-09-25 19:28:28.000000000 +0200
+++ pyinstaller-1.4b/source/linux/main.c    2010-09-19 20:10:09.000000000 +0200
@@ -28,6 +28,7 @@
 #include "launch.h"
 #include "getpath.h"
 #include <sys/wait.h>
+#include <signal.h>

 #ifdef FREEZE_EXCEPTIONS
 extern unsigned char M_exceptions[];
@@ -75,6 +76,12 @@
     int rc = 0;
     int pid;
     char *workpath = NULL;
+
+    // Ignore signals that should be handled by the python script
+    signal(SIGINT, SIG_IGN);
+    signal(SIGABRT, SIG_IGN);
+    signal(SIGTERM, SIG_IGN);
+
     /* atexit(cleanUp); */
 #ifdef FREEZE_EXCEPTIONS
     PyImport_FrozenModules = _PyImport_FrozenModules;
```

**Comment by dnephin** [comment]
We're hitting this issue very frequently with https://github.com/docker/compose (see #2055, #1040).  We've attached to work around the issue by implementing the signal handler in different ways, but in the end the issue is still there.

None of the issues occur when running the program from a pip install.

I'd be happy to test out any patches you might have to address the issue.

**Comment by dnephin** [comment]
Proposed a fix in #1822, maybe it's not right, but would appreciate any feedback about why it's not.

**Comment by htgoebel** [comment]
I tried looking into this, but my simple program (just `time.sleep(15)`) aborts correctly.

Does anybody have a small example which hits this but at least every second or third run?

**Comment by dnephin** [comment]
After looking into this again, I realized the pyinstaller bug was not exactly what I thought it was. The actual bug we're hitting is described here: https://github.com/docker/compose/issues/2251 and https://github.com/docker/compose/issues/1069.

When running the pyinstaller binary a `SIGINT` can cause the blocking `queue.get()` call to raise `thread.error`. I had worked around that by adding a `try/except thread.error: raise KeyboardInterupt()`, which was causing a different problem with our error handling.

I've been able to partially work around this issue by raising a different exception on `thread.error()`, but we still see cases where the signal handling does not work correctly when run from the pyinstaller binary (and we can't reproduce the issue when run from a `pip` install).

It's possible that it's related to the `queue.get()` exception still, but unfortunately I don't have a great minimal case to reproduce this.  I suspect you'd be able to reproduce the `thread.error()` issue with something like this:

```python
import Queue
queue = Queue.Queue()
queue.get(timeout=100)
```

and send the process a SIGINT when it's blocked.
